### PR TITLE
Kubelet Volume Manager Wait For Attach Detach Controller and Backoff on Error

### PR DESCRIPTION
**Pull request by saad-ali**
- Closes https://github.com/kubernetes/kubernetes/issues/27483
  - Modified Attach/Detach controller to report `Node.Status.AttachedVolumes` on successful attach (unique volume name along with device path).
  - Modified Kubelet Volume Manager wait for Attach/Detach controller to report success before proceeding with attach.
- Closes https://github.com/kubernetes/kubernetes/issues/27492
  - Implemented an exponential backoff mechanism for for volume manager and attach/detach controller to prevent operations (attach/detach/mount/unmount/wait for controller attach/etc) from executing back to back unchecked.
- Closes https://github.com/kubernetes/kubernetes/issues/26679
  - Modified volume `Attacher.WaitForAttach()` methods to uses the device path reported by the Attach/Detach controller in `Node.Status.AttachedVolumes` instead of calling out to cloud providers.

### CODE REVIEW

**Comment by erictune** [comment]
How is this coming?

**Comment by saad-ali** [comment]
> How is this coming?

Should have the PR ready to review today. Code review will likely take at least til Monday.

**Comment by saad-ali** [comment]
@caesarxuchao  Could you please review the "Add patch status to Node internalclientset" commit.
@jingxu97 Will review the volume code.

File: pkg/client/clientset_generated/internalclientset/typed/core/unversioned/node_expansion.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+// The NodeExpansion interface allows manually adding extra methods to the NodeInterface.
+type NodeExpansion interface {
+	// PatchStatus modifies the status of an existing node. It returns the copy
+	// of the node that the server returns, or an error.
+	PatchStatus(nodeName string, data []byte) (*api.Node, error)
+}
+
+// PatchStatus modifies the status of an existing node. It returns the copy of
+// the node that the server returns, or an error.
+func (c *nodes) PatchStatus(nodeName string, data []byte) (*api.Node, error) {
```
Comment (caesarxuchao) [inline pkg/client/clientset_generated/internalclientset/typed/core/unversioned/node_expansion.go]: FYI, https://github.com/kubernetes/kubernetes/pull/27293 is adding a Patch method to all clients, but it's not 1.3, and doesn't support subresource yet.

Comment (caesarxuchao) [inline pkg/client/clientset_generated/internalclientset/typed/core/unversioned/node_expansion.go]: The code LGTM for 1.3. I'll refactor it when #27293 is merged.

Comment (saad-ali) [inline pkg/client/clientset_generated/internalclientset/typed/core/unversioned/node_expansion.go]: Great, thanks @caesarxuchao!

File: pkg/util/goroutinemap/goroutinemap.go [hunk 2]
```diff
@@ -45,58 +56,133 @@ type GoRoutineMap interface {
...
+				operationName,
+				*err)
+		}
+	} else {
+		// Operation completed with error
+		existingOp := grm.operations[operationName]
+		if existingOp.durationBeforeRetry == 0 {
+			existingOp.durationBeforeRetry = initialDurationBeforeRetry
+		} else {
+			existingOp.durationBeforeRetry = 2 * existingOp.durationBeforeRetry
```
Comment (jingxu97) [inline pkg/util/goroutinemap/goroutinemap.go]: Do we want to set an upper bound for the durationBeforeRetry

File: pkg/util/goroutinemap/goroutinemap.go [hunk 3]
```diff
@@ -45,58 +56,133 @@ type GoRoutineMap interface {
...
+		// Operation completed without error, or exponentialBackOffOnError disabled
+		delete(grm.operations, operationName)
+		if *err != nil {
+			// Log error
+			glog.Errorf("operation for %q failed with: %v",
+				operationName,
+				*err)
+		}
+	} else {
+		// Operation completed with error
```
Comment (jingxu97) [inline pkg/util/goroutinemap/goroutinemap.go]: // Operation completed with error and exponentialBackOffOnerror Enabled

File: pkg/util/goroutinemap/goroutinemap.go [hunk 4]
```diff
@@ -45,58 +56,133 @@ type GoRoutineMap interface {
...
+			existingOp.durationBeforeRetry = initialDurationBeforeRetry
+		} else {
+			existingOp.durationBeforeRetry = 2 * existingOp.durationBeforeRetry
+		}
+		existingOp.lastError = *err
+		existingOp.lastErrorTime = time.Now()
+		existingOp.operationPending = false
+
+		grm.operations[operationName] = existingOp
+		if *err != nil {
```
Comment (jingxu97) [inline pkg/util/goroutinemap/goroutinemap.go]: err must be not nil here already, we can remove if

File: pkg/util/goroutinemap/goroutinemap.go [hunk 5]
```diff
@@ -45,58 +56,133 @@ type GoRoutineMap interface {
...
+			// Log error
+			glog.Errorf("Operation for %q failed. No retries permitted until %v (%v). error: %v",
```
Comment (jingxu97) [inline pkg/util/goroutinemap/goroutinemap.go]: Operation for %q failed. No retries permitted until %v (durationBeforeRetry %v). error: %v",

File: pkg/util/goroutinemap/goroutinemap_test.go [hunk 6]
```diff
@@ -106,9 +153,63 @@ func Test_NewGoRoutineMap_Positive_SecondOpAfterFirstPanics(t *testing.T) {
...
+	operation1 := generatePanicFunc()
+	err1 := grm.Run(operationName, operation1)
+	if err1 != nil {
+		t.Fatalf("NewGoRoutine failed. Expected: <no error> Actual: <%v>", err1)
+	}
+	operation2 := generateNoopFunc()
+
+	// Act
+	err2 := retryWithExponentialBackOff(
+		time.Duration(500*time.Millisecond), // Longer duration to accommodate for backoff
```
Comment (jingxu97) [inline pkg/util/goroutinemap/goroutinemap_test.go]: Can we put 500 as constant at the beginning? Also should we use a number larger than 500 to make sure it passed the retry period?

File: pkg/util/goroutinemap/goroutinemap_test.go [hunk 7]
```diff
@@ -174,44 +275,73 @@ func Test_NewGoRoutineMap_Positive_ThirdOpAfterFirstCompletes(t *testing.T) {
...
 	}
-}
 
-func generatePanicFunc() func() error {
-	return func() error {
-		panic("testing panic")
+	// Act
+	operation1DoneCh <- true // Force operation1 to complete
+	err3 := retryWithExponentialBackOff(
+		time.Duration(20*time.Millisecond),
```
Comment (jingxu97) [inline pkg/util/goroutinemap/goroutinemap_test.go]: put 20 at the beginning of the func as a constant?

File: pkg/util/goroutinemap/goroutinemap_test.go [hunk 8]
```diff
@@ -174,44 +275,73 @@ func Test_NewGoRoutineMap_Positive_ThirdOpAfterFirstCompletes(t *testing.T) {
...
+		func() (bool, error) {
+			err := grm.Run(operationName, operation3)
+			if err != nil {
+				t.Logf("Warning: NewGoRoutine failed. Expected: <no error> Actual: <%v>. Will retry.", err)
```
Comment (jingxu97) [inline pkg/util/goroutinemap/goroutinemap_test.go]: It is normal the run failed with error at the beginning, right? maybe move expected actual message, just say will retry?

File: pkg/volume/gce_pd/attacher.go [hunk 9]
```diff
@@ -89,7 +89,7 @@ func (attacher *gcePersistentDiskAttacher) Attach(spec *volume.Spec, hostName st
 	return path.Join(diskByIdPath, diskGooglePrefix+pdName), nil
 }
 
-func (attacher *gcePersistentDiskAttacher) WaitForAttach(spec *volume.Spec, timeout time.Duration) (string, error) {
+func (attacher *gcePersistentDiskAttacher) WaitForAttach(spec *volume.Spec, devicePath string, timeout time.Duration) (string, error) {
```
Comment (jingxu97) [inline pkg/volume/gce_pd/attacher.go]: So devicePath is not used for this function? Is that possible to use this instead of other information?

File: pkg/api/deep_copy_generated.go [hunk 10]
```diff
@@ -228,6 +229,12 @@ func DeepCopy_api_Affinity(in Affinity, out *Affinity, c *conversion.Cloner) err
 	return nil
 }
 
+func DeepCopy_api_AttachedVolume(in AttachedVolume, out *AttachedVolume, c *conversion.Cloner) error {
```
Comment (jingxu97) [inline pkg/api/deep_copy_generated.go]: what kind of error this function might return?

File: pkg/volume/aws_ebs/attacher.go [hunk 11]
```diff
@@ -41,38 +41,28 @@ func (plugin *awsElasticBlockStorePlugin) NewAttacher() (volume.Attacher, error)
...
 
 	awsCloud, err := getCloudProvider(attacher.host.GetCloudProvider())
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	attached, err := awsCloud.DiskIsAttached(volumeID, hostName)
+	// awsCloud.AttachDisk checks if disk is already attached to node and
+	// succeeds in that case, so no need to do that separately.
```
Comment (jingxu97) [inline pkg/volume/aws_ebs/attacher.go]: maybe rephrase this comment? I think you mean if err is not nil, that means disk attach failed, otherwise, attach success

File: pkg/volume/cinder/attacher.go [hunk 12]
```diff
@@ -79,38 +79,32 @@ func (attacher *cinderDiskAttacher) Attach(spec *volume.Spec, hostName string) e
 	if err == nil && attached {
 		// Volume is already attached to node.
 		glog.Infof("Attach operation is successful. volume %q is already attached to node %q.", volumeID, instanceid)
-		return nil
+	} else {
+		_, err = cloud.AttachDisk(instanceid, volumeID)
+		if err != nil {
```
Comment (jingxu97) [inline pkg/volume/cinder/attacher.go]: So if there is error, we just log it and continue to the next step cloud.GetAttachmentDiskPath?

File: pkg/volume/gce_pd/attacher.go [hunk 13]
```diff
@@ -79,18 +79,17 @@ func (attacher *gcePersistentDiskAttacher) Attach(spec *volume.Spec, hostName st
...
-		return err
+	} else {
+		if err := attacher.gceDisks.AttachDisk(pdName, hostName, readOnly); err != nil {
+			glog.Errorf("Error attaching PD %q to node %q: %+v", pdName, hostName, err)
+			return "", err
+		}
 	}
 
-	return nil
+	return path.Join(diskByIdPath, diskGooglePrefix+pdName), nil
```
Comment (jingxu97) [inline pkg/volume/gce_pd/attacher.go]: diskGooglePrefix + pdName (add space?)

File: pkg/volume/util/operationexecutor/operation_executor.go [hunk 14]
```diff
@@ -358,6 +386,20 @@ func (oe *operationExecutor) UnmountDevice(
 		string(deviceToDetach.VolumeName), unmountDeviceFunc)
 }
 
+func (oe *operationExecutor) VerifyControllerAttachedVolume(
```
Comment (jingxu97) [inline pkg/volume/util/operationexecutor/operation_executor.go]: Could we add comments for this func?

File: pkg/controller/volume/reconciler/reconciler.go [hunk 15]
```diff
@@ -88,10 +93,22 @@ func (rc *reconciler) reconciliationLoopFunc() func() {
...
-						glog.Infof("Started DetachVolume for volume %q to node %q", attachedVolume.VolumeName, attachedVolume.NodeName)
+						glog.Infof("Started DetachVolume for volume %q from node %q", attachedVolume.VolumeName, attachedVolume.NodeName)
+					}
+					if err != nil &&
+						!goroutinemap.IsAlreadyExists(err) &&
+						!goroutinemap.IsExponentialBackoff(err) {
+						// Ignore goroutinemap.IsAlreadyExists && goroutinemap.IsExponentialBackoff errors, they are expected.
+						// Log all other errors.
+						glog.Errorf(
+							"operationExecutor.DetachVolume failed to start for volume %q (spec.Name: %q) from node %q with err: %v",
```
Comment (jingxu97) [inline pkg/controller/volume/reconciler/reconciler.go]: I am thinking we can have some utility function to construct those similar glog which will reduce quite some lines from this big function. But we can do that later.

File: pkg/controller/volume/reconciler/reconciler_test.go [hunk 16]
```diff
@@ -32,6 +34,7 @@ import (
 const (
 	reconcilerLoopPeriod      time.Duration = 0 * time.Millisecond
```
Comment (jingxu97) [inline pkg/controller/volume/reconciler/reconciler_test.go]: we can to set this period as 0?

File: pkg/volume/util/operationexecutor/operation_executor.go [hunk 17]
```diff
@@ -793,15 +826,95 @@ func (oe *operationExecutor) generateUnmountDeviceFunc(
...
+				volumeToMount.VolumeSpec.Name(),
+				volumeToMount.PodName,
+				volumeToMount.Pod.UID)
+		}
+
+		for _, attachedVolume := range node.Status.VolumesAttached {
+			if attachedVolume.Name == volumeToMount.VolumeName {
+				addVolumeNodeErr := actualStateOfWorld.MarkVolumeAsAttached(
+					volumeToMount.VolumeSpec, nodeName, volumeToMount.DevicePath)
+				glog.Infof("Controller successfully attached volume %q (spec.Name: %q) pod %q (UID: %q)",
```
Comment (jingxu97) [inline pkg/volume/util/operationexecutor/operation_executor.go]: what if addVolumeNodeErr is not nil, is it still considered as successfully attached?

'

File: pkg/volume/util/operationexecutor/operation_executor.go [hunk 18]
```diff
@@ -793,15 +826,95 @@ func (oe *operationExecutor) generateUnmountDeviceFunc(
...
+		if !volumeToMount.PluginIsAttachable {
+			// If the volume does not implement the attacher interface, it is
+			// assumed to be attached and the the actual state of the world is
+			// updated accordingly.
+			addVolumeNodeErr := actualStateOfWorld.MarkVolumeAsAttached(
+				volumeToMount.VolumeSpec, nodeName, volumeToMount.DevicePath)
+			if addVolumeNodeErr != nil {
+				// On failure, return error. Caller will log and retry.
+				return fmt.Errorf(
+					"VerifyControllerAttachedVolume.MarkVolumeAsAttached failed for volume %q (spec.Name: %q) pod %q (UID: %q) with: %v.",
```
Comment (jingxu97) [inline pkg/volume/util/operationexecutor/operation_executor.go]: Just want to make sure, this error will be returned as func() error, or the second error?
'

File: pkg/volume/util/operationexecutor/operation_executor.go [hunk 19]
```diff
@@ -793,15 +826,95 @@ func (oe *operationExecutor) generateUnmountDeviceFunc(
...
+			}
+		}
+
+		// Volume not attached, return error. Caller will log and retry.
+		return fmt.Errorf("Volume %q (spec.Name: %q) pod %q (UID: %q) is not yet attached according to node status.",
+			volumeToMount.VolumeName,
+			volumeToMount.VolumeSpec.Name(),
+			volumeToMount.PodName,
+			volumeToMount.Pod.UID)
+	}, nil
```
Comment (jingxu97) [inline pkg/volume/util/operationexecutor/operation_executor.go]: does it mean the second error always be returned as nil?

File: pkg/util/goroutinemap/goroutinemap.go (hunk 2 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/goroutinemap.go]: Good idea. Done.

File: pkg/util/goroutinemap/goroutinemap.go (hunk 3 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/goroutinemap.go]: Done

File: pkg/util/goroutinemap/goroutinemap.go (hunk 4 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/goroutinemap.go]: Done.

File: pkg/util/goroutinemap/goroutinemap.go (hunk 5 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/goroutinemap.go]: Done

File: pkg/util/goroutinemap/goroutinemap_test.go (hunk 6 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/goroutinemap_test.go]: Moved to consts. Will leave it at 500, tested values higher and lower, and 500 results in the fastest test times.

File: pkg/util/goroutinemap/goroutinemap_test.go (hunk 7 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/goroutinemap_test.go]: Done.

File: pkg/util/goroutinemap/goroutinemap_test.go (hunk 8 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/goroutinemap_test.go]: Done.

File: pkg/volume/gce_pd/attacher.go (hunk 9 above)
Comment (saad-ali) [inline pkg/volume/gce_pd/attacher.go]: GCE PD generates it's own list of paths to check. It doesn't this without calling out to the cloud provider, so that's perfectly fine.

File: pkg/api/deep_copy_generated.go (hunk 10 above)
Comment (saad-ali) [inline pkg/api/deep_copy_generated.go]: This is generated code. No need to review.

File: pkg/volume/cinder/attacher.go (hunk 12 above)
Comment (saad-ali) [inline pkg/volume/cinder/attacher.go]: No, good catch. Fixed.

File: pkg/volume/gce_pd/attacher.go (hunk 13 above)
Comment (saad-ali) [inline pkg/volume/gce_pd/attacher.go]: gofmt does not permit spaces

File: pkg/controller/volume/reconciler/reconciler.go (hunk 15 above)
Comment (saad-ali) [inline pkg/controller/volume/reconciler/reconciler.go]: Agreed.

File: pkg/controller/volume/reconciler/reconciler_test.go (hunk 16 above)
Comment (saad-ali) [inline pkg/controller/volume/reconciler/reconciler_test.go]: Yep, meaning don't wait just execute as fast as possible.

File: pkg/volume/util/operationexecutor/operation_executor.go (hunk 17 above)
Comment (saad-ali) [inline pkg/volume/util/operationexecutor/operation_executor.go]: Yes, attach completed successfully and we must log that.

File: pkg/volume/util/operationexecutor/operation_executor.go (hunk 18 above)
Comment (saad-ali) [inline pkg/volume/util/operationexecutor/operation_executor.go]: Returned as the error to `func() error`

File: pkg/volume/util/operationexecutor/operation_executor.go (hunk 19 above)
Comment (saad-ali) [inline pkg/volume/util/operationexecutor/operation_executor.go]: In this case yes. If you look at the other `generate*` methods you'll see that this permits pre-processing before the method is created and an error to be returned in that case.

**Comment by saad-ali** [comment]
Thanks @jingxu97! Feedback addressed, PTAL

**Comment by jingxu97** [comment]
LGTM
